### ISSUE TITLE
Enquiry about band structure prediction

### ISSUE THREAD

**Issue by nukenadal** [QuantumLab-ZY/HamGNN#40]
Hi HamGNN developers,

I installed the package according to the instructions and built my first testing workflow, aiming to predict the band structure of a simple cubic CsPbI3 structure. I built a dataset that contains expanded and contracted cells and randomly perturbed structures (a total of 71 structures; the method is inspired by the original HamGNN publication). I then followed the instructions to train the Hamiltonian and band structure models, and tried to predict the band structure for a pristine relaxed structure that should be well-described by the dataset. 

I also checked that the predicted Hamiltonian values and band energy from the model are both accurate (as shown in the figures below). 

![Image](https://github.com/user-attachments/assets/86feff46-fdb3-479b-a6c5-8de4b851c2d1)
![Image](https://github.com/user-attachments/assets/cf72d72f-0b23-43ff-81ea-351aeadce79a)

However, in the prediction stage, I got a predicted band structure as:
![Image](https://github.com/user-attachments/assets/945b6dc0-4a29-4753-9ddc-1b618cbbf3ec)

The band_cal.yaml configuration is as follows:
```
nao_max: 26
graph_data_path: '/workspace/ham/predict/second/graph_data.npz'
hamiltonian_path: '/workspace/ham/cspbi3_1c_perturb/version_1/prediction_hamiltonian.npy'
nk: 80          # the number of k points
save_dir: '/workspace/ham/predict/second' # The directory to save the results
strcture_name: 'CsPbI3'  # The name of each cif file saved is strcture_name_idx.cif after band calculation
soc_switch: False
spin_colinear: False
auto_mode: False # If the auto_mode is used, users can omit providing k_path and label, as the program will automatically generate them based on the crystal symmetry.
k_path: [[0.0,0.0,0.5],[0.5,0.5,0.5],[0.0,0.5,0.5],[0.0,0.0,0.0],[0.5,0.5,0.5]]
label: ['$X$','$R$','$M$','$G$','$R$'] # The lable for each k points in K_path
```

I wonder if this somehow indicates an obvious error that can be easily tackled? Or maybe some of the parameters are incorrectly set. 

I truly appreciate your time and support on this matter. Thank you in advance for any further guidance you can offer. I look forward to your reply.

**Comment by QuantumLab-ZY** [comment]
Dear @nukenadal,

It seems the issue may not be related to diagonalization or network prediction. For the material CsPbI3, using random perturbations to construct a training set might not be suitable. CsPbI3 is a relatively soft material, and structural optimization can lead to significant relaxation. Therefore, small perturbations may not adequately sample its configuration space. I recommend using CsPbI3 configurations from molecular dynamics (MD) simulations for your training set instead.

Good luck,
Yang Zhong

**Comment by nukenadal** [comment]
Dear Yang Zhong,

Thank you so much for your prompt reply. 

I actually started from an MD-sampled dataset first, however the same issue persists. Then I turned to this perturbed method. You are right that for most perovskite, the soft bebaviour leads to spontaneous tilting, but the relaxed structure for inference I used are from the Materials Project which remains in a non-tilted cubic phase, so I assume that the perturbed structures should, in principle, cover this configuration. 

I was wondering if this issue may be caused by some other errors or mis-usage of the code?

Thank you so much in advance.

**Comment by QuantumLab-ZY** [comment]
Dear @nukenadal,

Based on your description, it seems there might be an issue with the network's prediction. I’ve previously trained HamGNN on this system, and the network performed well in fitting the band structures.

To better understand the problem, could you provide the following details?

1. What is the Mean Absolute Error (MAE) of the Hamiltonian matrix on the training set?
2. What is the MAE of the Hamiltonian matrix predicted by the network for this specific test structure?

Additionally, it would be helpful if you could share the following files for further debugging:

- `config.yaml` (the network configuration file)
- `.ckpt` (the trained model checkpoint)
- `POSCAR` (the structure file for the test case)

With this information, I can assist you more effectively in identifying and resolving the issue.

Best regards,  
Yang Zhong


> Dear Yang Zhong,
> 
> Thank you so much for your prompt reply.
> 
> I actually started from an MD-sampled dataset first, however the same issue persists. Then I turned to this perturbed method. You are right that for most perovskite, the soft bebaviour leads to spontaneous tilting, but the relaxed structure for inference I used are from the Materials Project which remains in a non-tilted cubic phase, so I assume that the perturbed structures should, in principle, cover this configuration.
> 
> I was wondering if this issue may be caused by some other errors or mis-usage of the code? Please let me know if additional info is required for better interpretation.
> 
> Thank you so much in advance.

**Comment by QuantumLab-ZY** [comment]
I cannot access the link you provided, but I have a suggestion regarding your issue.

The MAE for HamGNN on your training set being as low as 9.579e-09 seems unusually low. This might indicate that the model is inadvertently fitting against H0 (from `overlap.scfout`) instead of the intended Hamiltonian (H) from your DFT calculations. To address this issue, please verify that the Hamiltonian data from your DFT calculations has been correctly included in the graph_data.npz file. Please check your data packaging process and ensure that the Hamiltonian data is accurately included. 


> Thank you so much!
> 
> 1. The train MAE for the Hamiltonian matrix was 9.579e-09.
> 2. I am not really sure how to obtain the test error on this inferred structure. Could you remind me if it can be found somewhere in the inference output, or do I need to create a separate test set and run it with the model?
> 
> Additionally, please find the attached `config_hamiltonian.yaml` (for stage 1 training), `config_band.yaml` (stage 2), `checkpoint.ckpt` and `POSCAR` files (zipped, and passed through Google Drive due to max uploading file size in Github). Please let me know if you can access the files normally. `https://drive.google.com/file/d/1IwXcUthCMezMCvP7TFgCEtaCOvw1GRX0/view?usp=drive_link`
> 
> Best regards.

**Comment by QuantumLab-ZY** [comment]
When preparing the `graph_data.npz` for the training set, ensure that the `scfout_file_name` is set to the actual self-consistent Hamiltonian matrix generated by DFT calculations, such as `openmx.scfout`. Once your model is trained, you don't need to perform further DFT calculations. You can construct `graph_data.npz` using the `overlap.scfout` file from post-processing as the model's input.

**Comment by nukenadal** [comment]
Thank you so much for the explanation. 

So, in the case of future inference runs, for example with larger-supercell structures, do I need to perform OpenMX calculations to get the `openmx.scfout` file, then process it to get the `overlap.scfout` file, and ultimately use the overlap to construct `graph_data.npz` for inference? Or is there a way to by-pass the expensive DFT calculation to get the inference `graph_data.npz` file directly from structure files like POSCAR?

**Comment by QuantumLab-ZY** [comment]
As mentioned in the README, for model inference, you only need to run `openmx_postprocess` to generate the `overlap.scfout` file and use it for packaging `graph_data.npz` as model's input. At this moment, `scfout_file_name` should be set to `overlap.scfout`. This process does not involve executing any real DFT calculations.

![Image](https://github.com/user-attachments/assets/f3762523-7ba1-4b8e-a7e8-9b660a846da9)

**Comment by nukenadal** [comment]
Sorry for the misunderstanding of the README, maybe I have one more quick question: Could you remind me which file(s) in the directory will be read by the `openmx_postprocess` to produce `overlap.scfout`?

**Comment by QuantumLab-ZY** [comment]
`openmx_postprocess` is a modified version of OpenMX designed for analytically computing overlap matrices and Hamiltonian matrices H0. You can use `openmx_postprocess` in the same way you use `OpenMX`.

**Comment by nukenadal** [comment]
Thank you so much.

**Comment by nukenadal** [comment]
Hi @QuantumLab-ZY, 

I have a follow-up question from above. I redid the DFT calculations with a larger dataset and tried to repeat the workflow with the right settings we discussed above. However, during the graph generation stage, I got this error for every sub-folder to read:

`Failure of reading the scfout file (/workspace/ham/cspbi3_1c/300/overlap.scfout).` (where 300 is the last sub-folder)

Given that the `graph_data_gen.yaml` is as follows:

```
nao_max: 26
graph_data_save_path: '/workspace/ham/cspbi3_1c_graph'
read_openmx_path: '/home/HamGNN/openmx_postprocess/read_openmx'
max_SCF_skip: 200
scfout_paths: '/workspace/ham/cspbi3_1c/*' # The directory of the .scfout file calculated by openmx/openmx_postprocess, or a wildcard directory name to match multiple directories
dat_file_name: 'openmx.dat'
std_file_name: 'openmx.std' # Null if no openmx computation is performed
scfout_file_name: 'openmx.scfout' # If the openmx self-consistent Hamiltonian is not required as the target, "overlap.scfout" can be used instead.
soc_switch: False # generate graph_data.npz for SOC (True) or Non-SOC (False) Hamiltonian
# doping_charge: -1 # the background charge of the systems
```

Where I think I asked the code to read only the openmx output but not the overlap matrix from post-processing. I was wondering if I had done something wrong here in the settings? 

Many thanks.
